### ISSUE TITLE
dev-vboxguest.device "timed out"

### ISSUE THREAD

**Issue by domenkozar** [systemd/systemd#3423]
### Submission type
- [ x ] Bug report
### systemd version the issue has been seen with

systemd 229
+PAM +AUDIT -SELINUX +IMA +APPARMOR -SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN
### Used distribution

NixOS 16.03
### Description

We use an udev rule to teach systemd about `/dev/vboxguest` device so units can depend on it.

See the kernel rule at `https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/virtualisation/virtualbox-guest.nix#L80`

For some reason this doesn't work anymore since our 16.03 (when we upgraded systemd to 227 from 117) release:

```
# systemctl status dev-vboxguest.device
● dev-vboxguest.device
   Loaded: loaded
   Active: inactive (dead)

Jun 02 10:13:47 nixos systemd[1]: Found device /dev/vboxguest.
Jun 02 10:17:45 buggy systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start timed out.
Jun 02 10:17:45 buggy systemd[1]: Timed out waiting for device dev-vboxguest.device.
Jun 02 10:17:45 buggy systemd[1]: dev-vboxguest.device: Job dev-vboxguest.device/start failed with result 'timeout'.

# systemctl status systemd-udev-trigger.service
● systemd-udev-trigger.service - udev Coldplug all Devices
   Loaded: loaded (/nix/store/z9s789rbn3gbr2sykvbqg2pky1mx5kdm-systemd-229/example/systemd/system/systemd-udev-trigger.service; bad; vendor preset: enabled)
   Active: active (exited) since Thu 2016-06-02 12:23:13 UTC; 29min ago
     Docs: man:udev(7)
           man:systemd-udevd.service(8)
 Main PID: 1149 (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/systemd-udev-trigger.service

Jun 02 12:23:13 buggy systemd[1]: Starting udev Coldplug all Devices...
Jun 02 12:23:13 buggy systemd[1]: Started udev Coldplug all Devices.

ls -la /dev/vboxguest 
crw------- 1 root root 10, 58 Jun  2 12:23 /dev/vboxguest
# udevadm info -n /dev/vboxguest 

P: /devices/virtual/misc/vboxguest
N: vboxguest
E: DEVNAME=/dev/vboxguest
E: DEVPATH=/devices/virtual/misc/vboxguest
E: MAJOR=10
E: MINOR=58
E: PATH=/nix/store/15dyjdbrfchidxn3bgl2sndv0h0l6fjl-udev-path/bin:/nix/store/15dyjdbrfchidxn3bgl2sndv0h0l6fjl-udev-path/sbin
E: SUBSYSTEM=misc
E: TAGS=:systemd:
E: USEC_INITIALIZED=2823

# journalctl -b | grep vboxguest
Jun 02 10:04:45 nixos kernel: vboxguest: misc device minor 58, IRQ 20, I/O port d020, MMIO at 00000000f0000000 (size 0x400000)
Jun 02 10:04:45 nixos kernel: vboxguest: Successfully loaded version 5.0.4 (interface 0x00010004)
Jun 02 10:04:45 nixos systemd[1]: Expecting device dev-vboxguest.device...
Jun 02 10:04:46 nixos systemd[1]: Found device /dev/vboxguest.
```
- Kernel config: https://gist.github.com/84ef1f3c2a8c43a9d20b89b5b7564a90
- Software installed inside the guest: https://gist.github.com/e1fca205bb675b02c163c8f0ecc04963

Upstream bug report with a bit more details: https://github.com/NixOS/nixops/issues/391#issuecomment-223250485
### In case of bug report: Expected behaviour you didn't see

`dev-vboxguest.device` is started
### In case of bug report: Unexpected behaviour you saw

`dev-vboxguest.device` fails to start
### In case of bug report: Steps to reproduce the problem

1) Install VirtualBox with your Linux distro
2) `bash <(curl https://nixos.org/nix/install)`
3) Create a file `buggy.nix` with:

```
  {
    machine = {
        deployment.targetEnv = "virtualbox";
        deployment.virtualbox.headless = true;
        virtualisation.virtualbox.guest.enable = true;
    };
  }
```

4) `nixops create -d buggy ./buggy.nix`
5) `nixops deploy -d buggy`
6) `nixops ssh -d buggy machine -- systemctl status dev-vboxguest.device`

**Comment by edolstra** [comment]
It appears that `systemctl daemon-reload` (which NixOS runs when switching to a new configuration) causes systemd to forget all device units:

```
[root@machine:~]# systemctl --version
systemd 230
+PAM +AUDIT -SELINUX +IMA +APPARMOR -SMACK -SYSVINIT +UTMP -LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID -ELFUTILS +KMOD -IDN

[root@machine:~]# systemctl  | grep '\.device' | wc -l
23

[root@machine:~]# systemctl status sys-devices-virtual-misc-vboxguest.device
● sys-devices-virtual-misc-vboxguest.device - /sys/devices/virtual/misc/vboxguest
   Loaded: loaded
   Active: active (plugged) since Thu 2016-06-02 22:16:33 CEST; 27s ago
   Device: /sys/devices/virtual/misc/vboxguest

[root@machine:~]# systemctl daemon-reload

[root@machine:~]# systemctl status sys-devices-virtual-misc-vboxguest.device
● sys-devices-virtual-misc-vboxguest.device
   Loaded: loaded
   Active: inactive (dead)

[root@machine:~]# systemctl  | grep '\.device' | wc -l
0
```

I see this behaviour in VirtualBox but not on my host system. Maybe there is some virtualization check gone astray?

**Comment by edolstra** [comment]
Probably relevant log messages after `systemctl daemon-reload`:

```
Jun 02 22:32:20 machine systemd[1]: Reloading.
Jun 02 22:32:20 machine systemd[1]: Failed to enumerate devices: Invalid argument
```

**Comment by edolstra** [comment]
The `EINVAL` comes from `sd_device_new_from_subsystem_sysname` called with `subsystem="drivers"` and `sysname="vboxguest"`. This then hits:

```
                strscpy(subsys, sizeof(subsys), sysname);
                driver = strchr(subsys, ':');
                if (driver) {
                   ....
                } else
                        return -EINVAL;
```

**Comment by edolstra** [comment]
The cause was that we had a udev rule that didn't explicitly specify the subsystem:

```
        KERNEL=="vboxguest", TAG+="systemd"
```

which causes the creation of a tag file `/run/udev/tags/systemd/+drivers:vboxguest`, leading to `EINVAL` in `sd_device_new_from_subsystem_sysname`.

Changing this to

```
        SUBSYSTEM=="misc", KERNEL=="vboxguest", TAG+="systemd"
```

fixes the problem. However, maybe systemd should fill in the subsystem automatically, or complain about such rules.

**Comment by domenkozar** [comment]
I can confirm the fix. @poettering is this expected behavior? There could be at least be an error message.

**Comment by poettering** [comment]
no, this is a bug in systemd, PID 1 really should not choke on semi-borked udev rules, and udev really should complain anyway.

**Comment by poettering** [comment]
Fix waiting in #3608.

**Comment by poettering** [comment]
#3608 got applied, closing.
